k2/csrc/fsa_utils_inl.h

```diff
@@ -203,26 +199,228 @@ Array1<FloatType> GetForwardScores(FsaVec &fsas, Ragged<int32_t> &state_batches,
+      if (start_state_next_fsa - start_state > 0)
+        state_scores_data[start_state_next_fsa - 1] = -tot_scores_data[fsa_idx];
+    };
+    Eval(c, num_fsas, lambda_set_final_state_score);
+  } else {
+    // set the score of final state in each fsa to be 0
+    auto lambda_set_final_state_score =
+        [=] __host__ __device__(int32_t fsa_idx) {
+          int32_t start_state = fsa_row_splits1[fsa_idx],
+                  start_state_next_fsa = fsa_row_splits1[fsa_idx + 1];
+          if (start_state_next_fsa - start_state > 0)
+            state_scores_data[start_state_next_fsa - 1] = 0;
+        };
+    Eval(c, num_fsas, lambda_set_final_state_score);
+  }
+
+  // compute leaving arc numbers in each batch
+  Array1<int32_t> leaving_arc_nums(c, num_batches);
+  Array1<int32_t> leaving_arc_start_index(c, num_batches);
+  int32_t *leaving_arc_nums_data = leaving_arc_nums.Data();
+  int32_t *leaving_arc_start_index_data = leaving_arc_start_index.Data();
+  const int32_t *fsa_in_batch_row_splits =
+      leaving_arc_batches.RowSplits(1).Data();
+  const int32_t *state_in_batch_row_splits =
+      leaving_arc_batches.RowSplits(2).Data();
+  const int32_t *leaving_arc_row_splits =
+      leaving_arc_batches.RowSplits(3).Data();
+  auto lambda_set_leaving_arc_nums =
+      [=] __host__ __device__(int32_t batch_idx) {
+        int32_t first_state_idx_in_batch =
+                    state_in_batch_row_splits[batch_idx * num_fsas],
+                next_last_state_idx_in_batch =
+                    state_in_batch_row_splits[(batch_idx + 1) * num_fsas];
+        int32_t first_leaving_arc_idx =
+                    leaving_arc_row_splits[first_state_idx_in_batch],
+                next_last_leaving_arc_idx =
+                    leaving_arc_row_splits[next_last_state_idx_in_batch];
+        leaving_arc_start_index_data[batch_idx] = first_leaving_arc_idx;
+        leaving_arc_nums_data[batch_idx] =
+            next_last_leaving_arc_idx - first_leaving_arc_idx;
+      };
+  Eval(c, num_batches, lambda_set_leaving_arc_nums);
+
+  const int32_t *fsa_in_batch_row_ids = leaving_arc_batches.RowIds(1).Data();
+  const int32_t *state_in_batch_row_ids = leaving_arc_batches.RowIds(2).Data();
+  const int32_t *leaving_arc_row_ids = leaving_arc_batches.RowIds(3).Data();
+  const int32_t *leaving_arc_ids = leaving_arc_batches.values.Data();
+  const int32_t *states_data = state_batches.values.Data();
+  const Arc *arcs = fsas.values.Data();
+  Array1<FloatType> leaving_arc_score_values(
+      c, num_arcs);  // leaving arc_scores in batches
+  FloatType *arc_scores_data = leaving_arc_score_values.Data();
+  // convert leaving_arc_nums to cpu as we will access its elements in below
+  // Eval function for `lambda_set_leaving_arc_scores`
+  leaving_arc_nums = leaving_arc_nums.To(GetCpuContext());
+  const int32_t *cpu_leaving_arc_nums = leaving_arc_nums.Data();
+  Array1<int32_t> cpu_leaving_arc_start_index =
```

Comment:
since you already have this leaving_arc_start_index, it's not clear to me that you need leaving_arc_nums (could just take the difference in this).  Up to you though, no need to change.

Reply:
Removed leaving_arc_nums